File: ta4j-core/src/main/java/org/ta4j/core/rules/ChainRule.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.rules;

import java.util.LinkedList;
import java.util.Objects;
import java.util.Optional;

import org.ta4j.core.BarSeries;
import org.ta4j.core.Rule;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.rules.helper.ChainLink;

/**
 * A {@code ChainRule} has an initial rule that has to be satisfied before a
 * list of {@link ChainLink chain links} are evaluated. If the initial rule is
 * satisfied, each rule in {@link ChainRule#rulesInChain chain links} has to be
 * satisfied within a specified "number of bars (= threshold)".
 */
public class ChainRule extends AbstractRule {

    private final Rule initialRule;
    private final LinkedList<ChainLink> rulesInChain = new LinkedList<>();

    /** The backing series used to resolve the live begin index, or {@code null}. */
    private final BarSeries series;

    /**
     * @param initialRule the first rule that has to be satisfied before
     *                    {@link ChainLink} are evaluated
     * @param chainLinks  {@link ChainLink} that has to be satisfied after the
     *                    initial rule within their thresholds
     */
    public ChainRule(Rule initialRule, ChainLink... chainLinks) {
        this.initialRule = Objects.requireNonNull(initialRule, "initialRule cannot be null");
        Objects.requireNonNull(chainLinks, "chainLinks cannot be null");
        for (ChainLink chainLink : chainLinks) {
            this.rulesInChain.add(Objects.requireNonNull(chainLink, "chainLink cannot be null"));
        }
        this.series = findBarSeries(initialRule, rulesInChain);
    }

    private static BarSeries findBarSeries(Rule initialRule, LinkedList<ChainLink> rulesInChain) {
        return RuleCopies.findBarSeries(initialRule)
                .or(() -> rulesInChain.stream()
                        .map(chainLink -> RuleCopies.findBarSeries(chainLink.getRule()))
                        .flatMap(Optional::stream)
                        .findFirst())
                .orElse(null);
    }

    @Override
    public boolean isSatisfied(int index, TradingRecord tradingRecord) {
        // Resolve the retained begin index at evaluation time: on a rolling
        // series the constructor-time value goes stale as bars are evicted.
        final int beginIndex = series == null ? 0 : series.getBeginIndex();
        int lastRuleWasSatisfiedAfterBars = 0;
        int startIndex = index;

        if (!evaluateChildRule(initialRule, "initialRule", index, tradingRecord)) {
            if (isTraceEnabled()) {
                traceIsSatisfied(index, false, traceContext("initialRule", false));
            }
            return false;
        }

        int linkIndex = 0;
        for (ChainLink link : rulesInChain) {
            boolean satisfiedWithinThreshold = false;
            startIndex = startIndex - lastRuleWasSatisfiedAfterBars;
            lastRuleWasSatisfiedAfterBars = 0;

            for (int i = 0; i <= link.getThreshold(); i++) {
                int resultingIndex = startIndex - i;
                if (resultingIndex < beginIndex) {
                    break;
                }

                satisfiedWithinThreshold = evaluateChildRule(link.getRule(), "chainRule" + linkIndex, resultingIndex,
                        tradingRecord);

                if (satisfiedWithinThreshold) {
                    break;
                }

                lastRuleWasSatisfiedAfterBars++;
            }

            if (!satisfiedWithinThreshold) {
                if (isTraceEnabled()) {
                    traceIsSatisfied(index, false, traceContext("initialRule", true, "failedChainRule", linkIndex,
                            "threshold", link.getThreshold()));
                }
                return false;
            }
            linkIndex++;
        }

        if (isTraceEnabled()) {
            traceIsSatisfied(index, true, traceContext("initialRule", true, "chainRules", rulesInChain.size()));
        }
        return true;
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/rules/ChainRule.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.rules

import org.ta4j.core.*
import org.ta4j.core.rules.helper.ChainLink
import java.util.*

/**
 * A chainrule has an initial rule that has to be satisfied before chain links
 * are evaluated. If the initial rule is satisfied every rule of chain link has
 * to be satisfied within a specified amount of bars (threshold).
 *
 */
class ChainRule(private val initialRule: Rule, vararg chainLinks: ChainLink) : AbstractRule() {
    var rulesInChain = LinkedList<ChainLink>()

    /**
     * @param initialRule the first rule that has to be satisfied before
     * [ChainLink] are evaluated
     * @param chainLinks  [ChainLink] that has to be satisfied after the
     * inital rule within their thresholds
     */
    init {
        rulesInChain.addAll(chainLinks)
    }

    /** This rule uses the `tradingRecord`.  */
    override fun isSatisfied(index: Int, tradingRecord: TradingRecord?): Boolean {
        var lastRuleWasSatisfiedAfterBars = 0
        var startIndex = index
        if (!initialRule.isSatisfied(index, tradingRecord)) {
            traceIsSatisfied(index, false)
            return false
        }
        traceIsSatisfied(index, true)
        for (link in rulesInChain) {
            var satisfiedWithinThreshold = false
            startIndex = startIndex - lastRuleWasSatisfiedAfterBars
            lastRuleWasSatisfiedAfterBars = 0
            for (i in 0..link.threshold) {
                val resultingIndex = startIndex - i
                if (resultingIndex < 0) {
                    break
                }
                satisfiedWithinThreshold = link.rule.isSatisfied(resultingIndex, tradingRecord)
                if (satisfiedWithinThreshold == true) {
                    break
                }
                lastRuleWasSatisfiedAfterBars++
            }
            if (!satisfiedWithinThreshold) {
                traceIsSatisfied(index, false)
                return false
            }
        }
        traceIsSatisfied(index, true)
        return true
    }
}